File: web/src/api/privacy.ts
/** 隐私API只处理本人元数据和明确导出；不把密码、文件链接放入客户端持久存储。 */
import { objectValue, requestJson } from './client';
import { decodeProblem } from './problems';
import type { components } from './generated';
export type ExportRecord = components['schemas']['PersonalExport'];
export type ExportScope = ExportRecord['scope'];
export type TrashRecord = components['schemas']['TrashPage']['items'][number];
export type DeletionReceipt = components['schemas']['AccountDeletion'];
type Page<T> = { items: T[]; next_cursor: string | null };
/** value为文本，拒绝隐式转换。 */
function text(value: unknown): string { if (typeof value !== 'string') throw new Error('Invalid text'); return value; }
/** value为UUID，响应和路径共用校验，避免任意路径输入。 */
function uuid(value: unknown): string { const result = text(value); if (!/^[a-f\d]{8}-[a-f\d]{4}-[a-f\d]{4}-[a-f\d]{4}-[a-f\d]{12}$/i.test(result)) throw new Error('Invalid id'); return result.toLowerCase(); }
/** value为时间戳，只接受有效时间。 */
function date(value: unknown): string { const result = text(value); if (!Number.isFinite(Date.parse(result))) throw new Error('Invalid date'); return result; }
/** value为修订号，不允许精度丢失。 */
function count(value: unknown): number { if (typeof value !== 'number' || !Number.isSafeInteger(value) || value < 0) throw new Error('Invalid count'); return value; }
/** value与allowed为未知状态和固定允许值，不把新状态猜成完成。 */
function choice<T extends string>(value: unknown, allowed: readonly T[]): T { const result = text(value); if (!allowed.includes(result as T)) throw new Error('Invalid choice'); return result as T; }
/** value为当前导出任务公开状态，ready必须有下载时限。 */
function exportRecord(value: unknown): ExportRecord { const d = objectValue(value); const result = { id: uuid(d.id), scope: choice(d.scope, ['problems', 'learning', 'all_personal'] as const), status: choice(d.status, ['queued', 'running', 'ready', 'failed', 'expired'] as const), created_at: date(d.created_at), expires_at: d.expires_at === null ? null : date(d.expires_at), error_code: d.error_code === null ? null : text(d.error_code) }; if (result.status === 'ready' && !result.expires_at) throw new Error('Missing expiry'); return result; }
/** value为分页响应，decode只挑选当前资源白名单字段。 */
function page<T extends { id: string }>(value: unknown, decode: (value: unknown) => T): Page<T> { const d = objectValue(value); if (!Array.isArray(d.items) || d.items.length > 100) throw new Error('Invalid list'); const items = d.items.map(decode), next_cursor = d.next_cursor === null ? null : text(d.next_cursor); if (new Set(items.map(item => item.id)).size !== items.length || (next_cursor !== null && (!next_cursor || next_cursor.length > 4096))) throw new Error('Invalid page'); return { items, next_cursor }; }
/** cursor来自实际分页响应，不拼接私有信息到路径。 */
function query(cursor?: string | null): string { return '?' + new URLSearchParams(cursor ? { cursor } : {}); }
/** scope/password是本次明确输入，key只代表选择范围，不包含密码派生值。 */
export function requestExport(scope: ExportScope, password: string, key: string, signal?: AbortSignal) { return requestJson('/api/v1/me/exports', exportRecord, { method: 'POST', body: { scope, password }, idempotencyKey: key, signal }); }
/** cursor/signal限定当前身份和页面。 */
export function readExports(cursor?: string | null, signal?: AbortSignal) { return requestJson('/api/v1/me/exports' + query(cursor), value => page(value, exportRecord), { signal }); }
/** id是任务UUID；只接受同任务状态，不采用另一个文件的完成结果。 */
export function readExport(id: string, signal?: AbortSignal) { return requestJson(`/api/v1/me/exports/${uuid(id)}`, value => { const result = exportRecord(value); if (result.id !== id.toLowerCase()) throw new Error('Wrong export'); return result; }, { signal }); }
/** id/scope绑定本次明确下载；只接JSON附件，不跟随公开链接或渲染原始HTML。 */
export function downloadExport(id: string, scope: ExportScope, signal?: AbortSignal) { return requestJson(`/api/v1/me/exports/${uuid(id)}/download`, value => { const d = objectValue(value); if (d.schema_version !== 1 || uuid(d.export_id) !== id.toLowerCase() || d.scope !== scope) throw new Error('Wrong download'); date(d.generated_at); return new Blob([JSON.stringify(d, null, 2)], { type: 'application/json' }); }, { signal, timeoutMs: 60000 }); }
/** cursor/signal为本人回收站，不读取已删除问题的答案或知识卡正文。 */
export function readTrash(cursor?: string | null, signal?: AbortSignal) { return requestJson('/api/v1/me/trash' + query(cursor), value => page(value, (item): TrashRecord => { const d = objectValue(item); return { id: uuid(d.id), title: text(d.title), status: choice(d.status, ['deleted'] as const), revision: count(d.revision), deleted_at: date(d.deleted_at), purge_after: date(d.purge_after) }; }), { signal }); }
/** id/revision为当前问题快照，DELETE仅送回收站，不立即物理清理。 */
export function trashProblem(id: string, revision: number, signal?: AbortSignal) { return requestJson(`/api/v1/problems/${uuid(id)}`, value => { if (value !== undefined) throw new Error('Invalid deletion'); }, { method: 'DELETE', body: { expected_revision: revision }, signal }); }
/** id/revision为本人回收站条目，恢复归档不代表重新获得知识授权。 */
export function restoreProblem(id: string, revision: number, signal?: AbortSignal) { return requestJson(`/api/v1/problems/${uuid(id)}/restore`, value => { const result = decodeProblem(value); if (result.id !== id.toLowerCase() || result.status !== 'archived') throw new Error('Wrong restoration'); return result; }, { method: 'POST', body: { expected_revision: revision }, signal }); }
/** password/confirmation为本次明确注销，只有202确认才展示受理，不推定已清理。 */
export function requestDeletion(password: string, confirmation: string, signal?: AbortSignal) { return requestJson('/api/v1/me/deletion', (value): DeletionReceipt => { const d = objectValue(value); return { id: uuid(d.id), status: choice(d.status, ['pending'] as const), requested_at: date(d.requested_at), purge_after: date(d.purge_after) }; }, { method: 'POST', body: { password, confirmation }, signal }); }

File: server/operations/privacy.py
"""个人导出与软删除生命周期；身份锁保护再认证、修订和在途任务终结。"""
import hashlib
import json
from datetime import timedelta
from django.core import signing
from django.core.serializers.json import DjangoJSONEncoder
from django.db.models import Q
from django.utils import timezone
from django.utils.dateparse import parse_datetime
from access.context import owner_transaction
from accounts.services import AccountError
from problems.models import Problem, IdempotencyRecord
from problems.services import _current_user, _get, _public, _uuid, invalid, not_found, ProblemError
from runs.models import AnalysisRun, Job
from .models import PersonalExport


def digest(value):
    """value为许可或公开数据；稳定摘要避免持久保存完整权限记录。"""
    return hashlib.sha256(json.dumps(value, cls=DjangoJSONEncoder, sort_keys=True,
        ensure_ascii=False, separators=(',', ':')).encode()).hexdigest()


def reauthenticate(current, password):
    """current已持身份锁；password仅在内存中核验，不参与幂等摘要。"""
    if not isinstance(password, str) or not 1 <= len(password) <= 256 or not current.check_password(password):
        raise AccountError('INVALID_CREDENTIALS', '账号或密码不匹配或账号不可用', 401)


def export_public(row):
    """row为本人任务；到期即显示expired，不等待维护任务。"""
    expired = row.expires_at is not None and row.expires_at <= timezone.now()
    return {'id': str(row.pk), 'scope': row.scope, 'status': 'expired' if expired else row.status,
        'created_at': row.created_at, 'expires_at': row.expires_at, 'error_code': row.error_code}


def request_export(user, data, key):
    """user为会话身份；data为范围与瞬时密码；key绑定范围而非密码。"""
    if (not isinstance(data, dict) or set(data) != {'scope', 'password'}
            or data['scope'] not in ('problems', 'learning', 'all_personal')
            or not isinstance(key, str) or not key.strip() or len(key) > 128):
        raise invalid()
    with owner_transaction(user.pk):
        current = _current_user(user, lock=True)
        reauthenticate(current, data['password'])
        request_hash = digest({'scope': data['scope']})
        record = IdempotencyRecord.objects.filter(owner=current, method='POST',
            route_scope='/api/v1/me/exports', key=key).first()
        if record:
            if record.request_hash != request_hash:
                raise ProblemError('IDEMPOTENCY_CONFLICT', '幂等键已用于其他请求', 409)
            row = PersonalExport.objects.filter(owner=current, pk=record.resource_id).first()
            if row is None:
                raise not_found()
            return export_public(row)
        if PersonalExport.objects.filter(owner=current, status__in=['queued', 'running']).count() >= 3:
            raise ProblemError('EXPORT_LIMIT', '已有导出等待完成，请稍后重试', 429)
        row = PersonalExport.objects.create(owner=current, scope=data['scope'], auth_epoch=current.auth_epoch,
            access_revision=current.access_revision, deadline=timezone.now() + timedelta(minutes=15))
        result = export_public(row)
        IdempotencyRecord.objects.create(owner=current, method='POST', route_scope='/api/v1/me/exports',
            key=key, request_hash=request_hash, status='completed', response_status=202,
            response_payload=json.loads(json.dumps(result, cls=DjangoJSONEncoder)),
            resource_type='export', resource_id=row.pk, expires_at=timezone.now() + timedelta(hours=48))
        return result


def page(rows, query, user, kind):
    """rows为本人查询；query严格页大小，签名游标绑定身份epoch、范围和页大小。"""
    if not isinstance(query, dict) or not set(query) <= {'limit', 'cursor'}:
        raise invalid()
    limit = query.get('limit', 20)
    if type(limit) is not int or not 1 <= limit <= 100:
        raise invalid()
    binding = [str(user.pk), user.auth_epoch, kind, limit]
    field = 'deleted_at' if kind == 'trash' else 'created_at'
    if query.get('cursor'):
        try:
            value = signing.loads(query['cursor'], salt='privacy.page.v1', max_age=900)
            if value['binding'] != binding:
                raise ValueError('binding')
            timestamp = parse_datetime(value['time'])
            if timestamp is None or timezone.is_naive(timestamp):
                raise ValueError('timestamp')
            rows = rows.filter(Q(**{field + '__lt': timestamp}) | Q(**{field: timestamp,
                'id__lt': _uuid(value['id'], invalid)}))
        except (signing.BadSignature, KeyError, TypeError, ValueError):
            raise invalid() from None
    items = list(rows.order_by('-' + field, '-id')[:limit + 1])
    cursor = signing.dumps({'binding': binding, 'id': str(items[limit - 1].pk),
        'time': getattr(items[limit - 1], field).isoformat()},
        salt='privacy.page.v1') if len(items) > limit else None
    return items[:limit], cursor


def list_exports(user, query):
    """user/query为当前身份和页参数，仅列本人元数据。"""
    with owner_transaction(user.pk):
        current = _current_user(user)
        rows, cursor = page(PersonalExport.objects.filter(owner=current), query, current, 'exports')
        return {'items': [export_public(row) for row in rows], 'next_cursor': cursor}


def get_export(user, identifier):
    """user/identifier为当前身份与导出UUID。"""
    with owner_transaction(user.pk):
        current = _current_user(user)
        row = PersonalExport.objects.filter(owner=current, pk=_uuid(identifier, not_found)).first()
        if row is None:
            raise not_found()
        return export_public(row)


def cancel_runs(current, problem=None):
    """current已持锁；problem可限定父域，立即fence旧worker并正确释放或结算额度。"""
    from runs.queue import _terminal
    runs = AnalysisRun.objects.filter(owner=current, job__status__in=['queued', 'running', 'cancel_requested'])
    if problem is not None:
        runs = runs.filter(Q(problem=problem) | Q(learning_session__problem=problem))
    for run in runs.order_by('id'):
        job = Job.objects.select_for_update().get(owner=current, pk=run.job_id)
        _terminal(job, run, 'RUN_CANCELLED', timezone.now())


def revision(row, expected):
    """row为本人行；expected必须为精确非负整数。"""
    if type(expected) is not int or not 0 <= expected <= 9223372036854775806:
        raise invalid()
    if row.revision != expected:
        raise ProblemError('REVISION_CONFLICT', '问题已更新，请刷新后重试', 409, row.revision)


def delete_problem(user, identifier, expected):
    """user/identifier/expected为本人删除请求；授权丢失不阻止删除原始记录。"""
    with owner_transaction(user.pk):
        current = _current_user(user, lock=True)
        row = _get(current, _uuid(identifier, not_found), lock=True)
        revision(row, expected)
        row.status, row.deleted_at, row.revision = 'deleted', timezone.now(), row.revision + 1
        row.save(update_fields=['status', 'deleted_at', 'revision', 'updated_at'])
        cancel_runs(current, row)


def list_trash(user, query):
    """user/query为本人和分页；回收站只暴露元数据，不返回正文。"""
    with owner_transaction(user.pk):
        current = _current_user(user)
        rows, cursor = page(Problem.objects.filter(owner=current, status='deleted'), query, current, 'trash')
        return {'items': [{'id': str(row.pk), 'title': row.title, 'status': row.status,
            'revision': row.revision, 'deleted_at': row.deleted_at,
            'purge_after': row.deleted_at + timedelta(days=30)} for row in rows], 'next_cursor': cursor}


def restore_problem(user, identifier, expected):
    """user/identifier/expected为恢复请求；30天内归档恢复，不授予任何知识权限。"""
    with owner_transaction(user.pk):
        current = _current_user(user, lock=True)
        row = Problem.objects.select_for_update().filter(owner=current,
            pk=_uuid(identifier, not_found), status='deleted').first()
        if row is None:
            raise not_found()
        if row.deleted_at + timedelta(days=30) <= timezone.now():
            raise ProblemError('TRASH_EXPIRED', '恢复期限已过', 410)
        revision(row, expected)
        row.status, row.deleted_at, row.revision = 'archived', None, row.revision + 1
        row.save(update_fields=['status', 'deleted_at', 'revision', 'updated_at'])
        return _public(row, current)


def request_deletion(user, data):
    """user为会话本人，data为密码与固定确认文字；七天后由受控维护清除。"""
    from accounts.password_reset import invalidate_resets
    if not isinstance(data, dict) or set(data) != {'password', 'confirmation'} or data['confirmation'] != '删除我的账号':
        raise invalid()
    with owner_transaction(user.pk):
        current = _current_user(user, lock=True)
        reauthenticate(current, data['password'])
        now = timezone.now()
        current.status, current.deletion_requested_at = 'deletion_pending', now
        current.auth_epoch += 1
        current.access_revision += 1
        current.save(update_fields=['status', 'deletion_requested_at', 'auth_epoch', 'access_revision', 'updated_at'])
        invalidate_resets(current)
        cancel_runs(current)
        PersonalExport.objects.filter(owner=current, status__in=['queued', 'running', 'ready']).update(
            status='expired', expires_at=now, lease_token=None, lease_until=None, error_code='ACCESS_REVOKED')
        return {'id': str(current.pk), 'status': 'pending', 'requested_at': now,
            'purge_after': now + timedelta(days=7)}
